### PR TITLE
  Gpu miner

### PR DESCRIPTION
**Pull request by Gustav-Simonsson**
EDIT: branch moved to https://github.com/ethereum/go-ethereum/tree/gpu_miner

Please run:

```bash
geth gpuinfo
```

If it works you can then try with:
```bash
geth gpubench <device id>
```
and then add the GPU mining flag.

NOTE:
* Try geth gpuinfo first! This will tell you if it can find OpenCL and your GPU.
* You need to install a OpenCL installation. I've tested with AMDs.
* AMD cards on Linux may require root to run due to buggy GPU drivers

Ubuntu OpenCL instructions: http://wiki.tiker.net/OpenCLHowTo

### CODE REVIEW

**Comment by Gustav-Simonsson** [comment]
DO NOT MERGE YET! WIP

**Comment by obscuren** [comment]
```
# github.com/Gustav-Simonsson/go-opencl/cl
ld: library not found for -lOpenCL
clang: error: linker command failed with exit code 1 (use -v to see invocation)
godep: go exit status 2
```

File: Godeps/_workspace/src/github.com/Gustav-Simonsson/go-opencl/cl/cl.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+package cl
+
+// #ifdef __APPLE__
+// #include "OpenCL/opencl.h"
+// #else
+// #include "cl.h"
+// #endif
+// #cgo CFLAGS: -I/opt/opencl-headers/include
+// #cgo darwin LDFLAGS: -framework OpenCL
+// #cgo LDFLAGS: -lOpenCL
```
Comment (obscuren) [inline Godeps/_workspace/src/github.com/Gustav-Simonsson/go-opencl/cl/cl.go]: @Gustav-Simonsson You need to change this in the base repo to something like `#cgo linux LDFLAGS: -LOpenCL`

**Comment by kobigurk** [comment]
gpubench crashes for me :  https://gist.github.com/kobigurk/be4f11d87ad98f81e286

**Comment by zelig** [comment]
@Gustav-Simonsson it only works with `sudo` for me.

Although I run it with `--gpuchunks` option I get a warning that the dagchunk is larger than mem alloc size:

```
Warning: DAG size (1082130304) larger than device max memory allocation size (1073741824).
You may have to run with --opencl-mem-chunking
```

Note also the option `--opencl-mem-chunking` in the message should probably read `--gpuchunks`
and should be written on stderr instead of stdout.

Then after `Generating cache for epoch 1`, it crashes
 https://gist.github.com/zelig/f38d32012886836c9db5

**Comment by etherninja** [comment]
Complied after adding #include "CL/cl.h" to files. works but ignores secondary cards

/miner/go-ethereum/build/bin/geth --minegpu --minerthreads 2 --autodag --datadir....

**Comment by obscuren** [comment]
Getting the following error `OpenCL init error:  Device id not found: <nil> See device ids with: geth gpuinfo`

```
=============================================
============ OpenCL Device Info =============
=============================================
Platform id              0
Platform Name            AMD Accelerated Parallel Processing
Platform Vendor          Advanced Micro Devices, Inc.
Platform Version         OpenCL 2.0 AMD-APP (1729.3)
Platform Extensions      cl_khr_icd cl_amd_event_callback cl_amd_offline_devices
Platform Profile         FULL_PROFILE

Device OpenCL id         0
Device id for mining     0
Device Name              Tahiti
Vendor                   Advanced Micro Devices, Inc.
Version                  OpenCL 1.2 AMD-APP (1729.3)
Driver version           1729.3 (VM)
Address bits             32
Max clock freq           1020
Global mem size          3182428160
Max constant buffer size 65536
Max mem alloc size       2205155328
Max compute units        32
Max work group size      256
Max work item sizes      [256 256 256]
=============================================
Device OpenCL id         0
Device id for mining     1
Device Name              Tahiti
Vendor                   Advanced Micro Devices, Inc.
Version                  OpenCL 1.2 AMD-APP (1729.3)
Driver version           1729.3 (VM)
Address bits             32
Max clock freq           1020
Global mem size          3182428160
Max constant buffer size 65536
Max mem alloc size       2205155328
Max compute units        32
Max work group size      256
Max work item sizes      [256 256 256]
=============================================
Device OpenCL id         0
Device id for mining     2
Device Name              Tahiti
Vendor                   Advanced Micro Devices, Inc.
Version                  OpenCL 1.2 AMD-APP (1729.3)
Driver version           1729.3 (VM)
Address bits             32
Max clock freq           1020
Global mem size          3182428160
Max constant buffer size 65536
Max mem alloc size       2205155328
Max compute units        32
Max work group size      256
Max work item sizes      [256 256 256]
=============================================
Device OpenCL id         0
Device id for mining     3
Device Name              Tahiti
Vendor                   Advanced Micro Devices, Inc.
Version                  OpenCL 1.2 AMD-APP (1729.3)
Driver version           1729.3 (VM)
Address bits             32
Max clock freq           1020
Global mem size          3182428160
Max constant buffer size 65536
Max mem alloc size       2205155328
Max compute units        32
Max work group size      256
Max work item sizes      [256 256 256]
=============================================
Found 4 devices. Benchmark first GPU:       geth gpubench 0
Mine using all GPUs:                        geth --minegpu 0,1,2,3
You may also need to enable chunking:            --gpuchunks
```

**Comment by Gustav-Simonsson** [comment]
@obscuren try re-fetching branch, pushed a fix to gpubench, should work as expected now.

**Pull request closed without merging**